### PR TITLE
zypper_package is not idempotent on SUSE

### PR DESCRIPTION
**Pull request by ashwin-msys**
- Fixed the issue by handling the exception which raises if SUSE system is not able to search a particular package.

Signed-off-by: Ashwin <avarma@msystechnologies.com>

## Description
Fixed the issue by handling the exception which raises if SUSE system is not able to search a particular package.

## Related Issue
https://github.com/chef/chef/issues/11357
https://github.com/chef/customer-bugs/issues/335

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/provider/package/zypper.rb [hunk 1]
```diff
@@ -97,7 +97,11 @@ def resolve_current_version(package_name)
 
         def resolve_available_version(package_name, new_version)
           search_string = new_version.nil? ? package_name : "#{package_name}=#{new_version}"
-          so = shell_out!("zypper", "--non-interactive", "search", "-s", "--provides", "--match-exact", "--type=package", search_string)
+          so = begin
+                shell_out!("zypper", "--non-interactive", "search", "-s", "--provides", "--match-exact", "--type=package", search_string)
+               rescue Exception => e
+                 Mixlib::ShellOut.new
+              end
```
Comment (lamont-granquist) [inline lib/chef/provider/package/zypper.rb:104]: that just needs to be converted to `shell_out` instead of `shell_out!` which will probably fix it

**Review by lamont-granquist — CHANGES_REQUESTED** [review]

File: lib/chef/provider/package/zypper.rb (hunk 1 above)
Comment (ashwin-msys) [reply lib/chef/provider/package/zypper.rb]: Thanks for the suggestion @lamont-granquist. I've made the changes.

**Comment by lamont-granquist** [comment]
There needs to be functional tests added here:

https://github.com/chef/chef/blob/master/spec/functional/resource/zypper_package_spec.rb

File: spec/unit/provider/package/zypper_spec.rb [hunk 2]
```diff
@@ -248,6 +248,13 @@ def shell_out_expectation!(*command, **options)
         )
         provider.remove_package(["emacs"], ["1.0"])
       end
+
+      it "should run zypper remove for non existing package" do
+        shell_out_expectation!(
+          "zypper", "--non-interactive", "remove", "noname=1.0"
+        )
```
Comment (lamont-granquist) [inline spec/unit/provider/package/zypper_spec.rb:255]: this expectation doesn't have anything to do with the changed code so its difficult to see how this is relevant and looks like its not really testing the behavior.

Comment (marcparadise) [reply spec/unit/provider/package/zypper_spec.rb:255]: Maybe if the description is changed, "should run zypper remove for non-existent package without raising an error"?

**Review by marcparadise — CHANGES_REQUESTED** [review]

**Comment by marcparadise** [comment]
It would also be good to include a test that `resolve_available_version` returns nil when it can't find the package.

**Comment by tas50** [comment]
@ashwin-msys Can you update the implementation based on the feedback here?

File: spec/unit/provider/package/zypper_spec.rb (hunk 2 above)
Comment (ashwin-msys) [reply spec/unit/provider/package/zypper_spec.rb:255]: Hi @marcparadise Thanks for the suggestion. I've updated the spec description.

**Comment by ashwin-msys** [comment]
@lamont-granquist Yes, I already checked for the functional specs. So its already there in the file you suggested.
You can check them https://github.com/chef/chef/blob/master/spec/functional/resource/zypper_package_spec.rb on line number: 151

Please let me know if we need to add any spec more specifically. 
Thank you!

**Comment by ashwin-msys** [comment]
> It would also be good to include a test that `resolve_available_version` returns nil when it can't find the package.

@marcparadise I am on it. My suse instance was terminated. So getting it ready and will add the specific spec as well.

**Comment by ashwin-msys** [comment]
> It would also be good to include a test that `resolve_available_version` returns nil when it can't find the package.

Hi @marcparadise , I have added the requested spec here for #resolve_available_version method. Please check

**Comment by lamont-granquist** [comment]
no you need to add a test to the functional specs to remove and rpm that doesn't exist so instead of `:remove`'ing the chef_rpm package, you need to do that to a "this_package_should_never_exist_cheftest" package or something like that.

**Comment by ashwin-msys** [comment]
> no you need to add a test to the functional specs to remove and rpm that doesn't exist so instead of `:remove`'ing the chef_rpm package, you need to do that to a "this_package_should_never_exist_cheftest" package or something like that.

Hi @lamont-granquist I've added the functional specs here. Please check :)

**Review by jaymzh — COMMENTED** [review]
So the problem here is we stop failing on the check-if-the-package-is-installed, but then we remove packages that are not there (won't fail, but not idempotent), as is proven by the added test.

We probably want simple idempotency checks on the top-level entry
methods on line 41-64. This should be easy there is an `is_installed`.

Then update all the test to test for idempotent behavior, not repetative behavior.

**Review by jaymzh — CHANGES_REQUESTED** [review]
Requesting changes for my last comment
